var/spack/repos/builtin/packages/gdal/package.py

```diff
@@ -629,7 +629,6 @@ def configure_args(self):
             self.with_or_without("hdf4", package="hdf"),
             self.with_or_without("hdf5", package="hdf5"),
             self.with_or_without("hdfs", package="hadoop"),
-            self.with_or_without("libiconv-prefix", variant="iconv", package="iconv"),
```

Comment:
What's the build error you see without this?

Reply:
If `iconv` is provided by `libc` and is an external, then this adds `/usr` as a high-priority prefix.

Reply:
That sounds like it isn't a bug with libc but a bug with external packages. Shouldn't you instead check where iconv is installed and filter out system paths? `spack.util.environment` has `is_system_path` and `filter_system_paths` functions to handle this kind of stuff.

Reply:
I need to double-check, but I'm moderately sure that using `--libiconv-prefix` tries to add `-liconv` to the library list  which will fail if we're getting it from `libc`. The fouling of the library search path is secondary to that.

With reference to, "Shouldn't you instead check where iconv is installed and filter out system paths?" however: if "you" is, the package maintainer, the unfortunate answer to that is, "no, because nobody else does." As a release manager who configures a certain set of externals all the time, I'm forever running into packages where some aspect of the recipe: library search paths, `PATH`, or other environment settings gets tripped up by using externals. It's left to individual Spack users like me to trip over the problem, fix it in every package that has the problem and submit the PR. If there were a way to have the CI test recipe builds with dependencies configured as externals in addition to having them installed in the same Spack setup, that might help. Until then however, the burden falls on the first unfortunate soul to want to build against _that_ dependency as an external and gets to watch everything explode.

Reply:
By "you" I mean the person writing or maintaining the `package.py` build recipes. From your initial reply, it sounded like instead of checking whether iconv is libc, we should instead check to see if iconv is installed in a system directory where many other things are installed.

We automatically filter system deps most of the time (anything automatically added to the compiler wrappers, for example). It's hard to automatically filter them the rest of the time though. The best we can do automatically would be to ensure that all link/include flags are sorted with system locations last. Can't remember if @scheibelp or @becker33 or @haampie ever managed to get that working, or who was working on that. I agree that it's annoying to have to do this on a package-by-package basis.

Reply:
> By "you" I mean the person writing or maintaining the `package.py` build recipes. From your initial reply, it sounded like instead of checking whether iconv is libc, we should instead check to see if iconv is installed in a system directory where many other things are installed.

No, sorry.

> We automatically filter system deps most of the time (anything automatically added to the compiler wrappers, for example). It's hard to automatically filter them the rest of the time though. The best we can do automatically would be to ensure that all link/include flags are sorted with system locations last. Can't remember if @scheibelp or @becker33 or @haampie ever managed to get that working, or who was working on that. I agree that it's annoying to have to do this on a package-by-package basis.

We already have environment utilities provided by me to do this:https://github.com/spack/spack/blob/9b3c4e0696a9f90ac0870c32b0ef9dad1b529583/lib/spack/spack/util/environment.py#L76-L80 and https://github.com/spack/spack/blob/9b3c4e0696a9f90ac0870c32b0ef9dad1b529583/lib/spack/spack/util/environment.py#L330-L338 but that has to be used on a package-by-package basis. I'm not sure how easy that would be to integrate into the wrapper's operations.

Reply:
It's already used

Reply:
@greenc-FNAL I'll note that our compiler wrappers reorder `-L` etc. so even if the configure script introduces a mixed order of system/custom paths, our wrapper should disentangle that. It would be useful to see an example failed installation and specifically the corresponding modified compiler invocations you can see in `spack-cc..out` if you run the install with `-d` like `spack -d install ...`.